designs/2021-init-command-eslint-cli/README.md

```diff
@@ -0,0 +1,153 @@
+We will make use of tools like `npm` to simply run `npm init @eslint/config` or using `npx` to run `npx  @eslint/create-config` or using `yarn` to run `yarn create @eslint/config` single time for a project instead of having it in the core project.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+The implementation is mainly migrating the code from main repo to `@eslint/create-config`.
+
+The approach would be following these steps
+
+### 1. Remove eslint auto-config
+
+- Remove auto-config(`eslint/lib/init/autoconfig.js`), and its tests(`eslint/tests/lib/init/autoconfig.js`).
+- Remove dependency `progress`, as it's no longer used in production.
+
+### 2. Move `eslint --init` related files to a separate repo
+
+- Move the `eslint/lib/init/*` to the new repo's `lib/*` directory
+- Add the following `dependencies` in `package.json`
+
+  - `enquirer`
+  - `progress`
+  - `semver`
+  - `lodash`
+  - `json-stable-stringify-without-jsonify`
+  - `cross-spawn`
+  - `debug`
+  - `@eslint/eslintrc`
+  - `eslint` (to be removed)
+  - `espree` (to be removed)
+
+- for all those modules coming inside `../**/*`, it would be replaced using `eslint/**/*`
+- For `tests`, move the `tests/lib/init` to new repo's `tests/lib/`
+- Add the following `devDependencies` in `package.json`
+
+  - `chai`
+  - `sinon`
+  - `js-yaml`
+  - `proxyquire`
+  - `shelljs`
+
+It was almost done: [aladdin-add/eslint-create](https://github.com/aladdin-add/eslint-create).
+I will make a PR once the official repo is created later.
+
+### 3. Remove the usage of eslint & espree
+
+eslint was used:
+https://github.com/aladdin-add/eslint-create/blob/c27509cbbaea7c846fd9d7a373feece5254ac8d8/lib/config-file.js#L86
+
+It can be removed by using the local installed eslint, and switch to the new `Linter` api.
+
+espree was used:
+https://github.com/aladdin-add/eslint-create/blob/c27509cbbaea7c846fd9d7a373feece5254ac8d8/lib/config-initializer.js#L161
+
+It can be removed after https://github.com/eslint/espree/issues/495 get landed.
+
+### 4. Release the new package
+
+### 5. Update eslint repo
+
+- Remove `eslint --init` related files.
+- the following dependencies can be removed in eslint repo:
+  - enquirer
+  - semver
+- Whenever `--init` command is being used, show a warning(e.g. "You can also run this command directly using 'npm init @eslint/config'") and run `npm init @eslint/config` using `child_process` of the native node modules.
```

Comment:
👍 good point! will update later.

Reply:
Let's wait for more opinions first. @nzakas, @btmills what do you think about this?

Reply:
Doesn't that issue only apply when using `yarn`? If we run `npm init @eslint/config`, aren't we okay?

Reply:
this is talking about the  eslint cli(`eslint --init`), not the new `eslint/create-config`.

It will be an issue when using some different package manager than npm(yarn/pnpm/...).

Reply:
> Doesn't that issue only apply when using `yarn`? If we run `npm init @eslint/config`, aren't we okay?

It should be okay with `npm` since we will probably exit the process right after `npm init @eslint/config` finishes, but I think it's always better to avoid initiating changes to `node_modules` from a module that is in the same `node_modules` (`eslint --init` could refer to a locally installed ESLint) because further execution becomes unpredictable. Paths in `node_modules` may have changed, some not-yet-loaded modules may have been overwritten with different versions of themselves, etc. Yarn 2 detects this and aborts the process with an error message.

If that would be an equally non-breaking change, my suggestion is to minimize possible issues with any package manager by just printing the command instead of running it automatically.

Reply:
> it's always better to avoid initiating changes to node_modules from a module that is in the same node_modules

That makes sense. I'm fine just printing the command instead of running it.

Reply:
Printing the command is essentially causing `eslint --init` to error out vs. transparently doing what it does right now. That's fine, but I'm not sure we are buying much safety vs. what we already have today.

Reply:
currently it's ok, as `eslint --init` only officially supports npm.

But we have a plan to support yarn as well. `npm init @eslint/config` assumes users are using npm - that's not always true. it might cause some issues for yarn users, e.g. yarn.lock not updated.

or just table it, and consider this when support yarn?

Reply:
The command we use to run init can be different than the package manager init uses to install dependencies. We can hard code `npm init @eslint/create-config` but still detect and use e.g. `yarn` to install dependencies.

Reply:
that also works. though I think better to recommend yarn users run `yarn create @eslint/config`. :)

Reply:
`@eslint/create-config` will `require()` local ESLint to format the `.eslintrc.js` config file. Since yarn 2 doesn't create `node_modules`, I _think_ that `@eslint/create-config` when run through `npm init` won't be able to `require()` any locally installed packages.

Reply:
To add to my previous comment, `eslint --init` doesn't officially support yarn yet, but `yarn eslint --init` currently works fine when provided answers don't require installing dependencies. I'd expect the [formatting step](https://github.com/eslint/eslint/blob/8490fb42e559ef0b3c34ac60be4e05e0d879a9cb/lib/init/config-file.js#L86-L94) to fail when user runs `yarn eslint --init` and then we run `npm init @eslint/config` as a child process.

Reply:
Given that we are in final commenting and there’s a lot of “I think” comments that are best answered by implementing the package and testing it, I move that we approve this RFC and hold further speculation until there is a prototype to play with.

Reply:
Agreed. We'll certainly publish the new package first, and then we can experiment.